Remember debugger size and position

hernanmd opened the pull request:
Add lastKnownExtent and lastKnownPosition to preserve the last extent and position for the next debug request window.
Add a setting to enable this behavior (enabled by default after this PR).

Commit:
Add lastKnownExtent and lastKnownPosition to preserve the last extent and position for the next debug request window.
Add a setting to enable this behavior (enabled by default after this PR).

StevenCostiou commented:
Thank you Hernan.

However I would not like to do it like that, we spent a lot of time removing similar code from the debugger.
It solves the problem (I think) but it only add unnecessary complexity to the debugger, using globals etc.

I would like to find time to discuss the solution, for instance to move all settings of the debugger into a single "debugger setting" object instead of class variables.

Please also come to the debugging team to discuss before implementing without us knowing, because we can share a bit of vision and technical constraints and objectives.

Commit:
Cleaner StDebugger implementation by moving the settings methods to a new class StDebuggerSettings.

Commit:
Finish cleanup

StevenCostiou commented:
@adri09070 can you have a look?

hernanmd commented:
> Thank you Hernan.
> 
> However I would not like to do it like that, we spent a lot of time removing similar code from the debugger. It solves the problem (I think) but it only add unnecessary complexity to the debugger, using globals etc.
> 
> I would like to find time to discuss the solution, for instance to move all settings of the debugger into a single "debugger setting" object instead of class variables.
> 
> Please also come to the debugging team to discuss before implementing without us knowing, because we can share a bit of vision and technical constraints and objectives.

Hi Steven,

I've updated the PR with a cleaner StDebugger "remember position" implementation by moving the settings methods to a new class, StDebuggerSettings. Please let me know of any appropriate enhancements.

Commit:
Use #defaultExtent in settings class, with scaledByDisplayScaleFactor.

Commit:
Avoid setting the position and extent when the window is not built yet.

hernanmd commented:
I've fixed some additional issues, so the CI should pass and can be reviewed "safely"

StevenCostiou commented on src/NewTools-Debugger/StDebugger.class.st:
```diff
@@ -498,6 +498,12 @@ StDebugger >> debuggerInspectorModelClass [
 ]
 
+{ #category : 'initialization' }
+StDebugger >> debuggerSettings [
```
the method name should be simply "settings"

StevenCostiou replied:
it should also be class side

hernanmd replied:
You mean?

```smalltalk
initializeWindow: aWindowPresenter

 "blabla"
self class settings initializeSettings: aWindowPresenter.

```

StevenCostiou replied:
Rather something like `self settings setWindowPosition: aWindowPresenter` and `self settings` is implemented as `^self class settings`.

The parameter to a method named `initializeSettings:` should not be a presenter but a setting or something else.

hernanmd replied:
Done

StevenCostiou commented on src/NewTools-Debugger/StDebugger.class.st:
```diff
@@ -755,6 +761,12 @@ StDebugger >> initializePresenters [
 ]
 
+{ #category : 'initialization' }
+StDebugger >> initializeSettings: aWindowPresenter [
+
+	self debuggerSettings initializeSettings: aWindowPresenter
+]
```
the call to the settings should be inlined in the method needing it.
here there is no reason that the parameter would be a window presenter

StevenCostiou replied:
hence I would remove that method entirely

hernanmd replied:
Done

StevenCostiou commented on src/NewTools-Debugger/StDebugger.class.st:
```diff
@@ -814,9 +826,11 @@ StDebugger >> initializeWindow: aWindowPresenter [
 		whenClosedDo: [ self clear ].
-
+		
+	self initializeSettings: aWindowPresenter.
```
I would inline the window extent setting here (by calling the settings)

hernanmd replied:
Done

StevenCostiou commented on src/NewTools-Debugger/StDebuggerSettings.class.st:
```diff
@@ -0,0 +1,128 @@
+]
+
+{ #category : 'initialization' }
+StDebuggerSettings class >> initializeSettings: aWindowPresenter [
+
+	self preserveWindowPosition ifFalse: [ ^ self ].
+	aWindowPresenter window ifNil: [ ^ self ].
+	aWindowPresenter whenClosedDo: [
+		self
+			lastKnownPosition: aWindowPresenter window position;
+			lastKnownExtent: aWindowPresenter window extent ]
+]
```
this method should not be called "initializeSettings:" but rather "configure debugger window" or something similar.

hernanmd replied:
For compatibility with Spec 2, I renamed it as `initializeWindow:`

StevenCostiou replied:
I don't know if it is an initialization really, as the usage is in the end to capture in the settings the last known position of the window.

I also think that there should not be any compatbility with spec, perhaps settings should have an explicit "setting" api.

I didn't catch it at start, but any chance of getting rid of the nil check?
Is it possible that a closing window presenter has no window?
I know sometimes spec does strange things and we can't do anything but I like to try.

StevenCostiou replied:
thanks for the hard work :)

StevenCostiou reviewed (changes requested):
In addition to my comments, we need tests ensuring the new introduced behavior.

Commit:
Modifications implemented as suggested by @StevenCostiou

Commit:
Add preserve window position test.
Add convenience settings method.
Also settings should be on class side.
Fix tearDown issue while debugging the debugger window which lost its session.

Commit:
Add preserve extent test.

Commit:
Use more descriptive accessors preserveWindowPositionAndExtent.
Use more descriptive variable name PreserveWindowPositionAndExtent
Finish debugger position and extent tests.

hernanmd commented:
> In addition to my comments, we need tests ensuring the new introduced behavior.

Done in the last PR commit.

Commit:
Fix for testClose

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,56 @@ StDebuggerTest >> testNewDebuggerContextFor [
+	dbg window window extent: newDbgExtent.
+	"New extent is saved when the window is closed"
+	dbg close.
+	self tearDown.
+
+	"Now we open the debugger again"
+	self setUp.
+	dbg := (self debuggerOn: session) open.
+	"Check the debugger window is set to the new extent"
+	self assert: dbg window window extent equals: newDbgExtent.
+	"Set back the last extent regardless of this test"
+	dbg window window extent: lastUserExtent.
+
+]
+
+{ #category : 'tests - context inspector' }
+StDebuggerTest >> testPreserveLastKnownPosition [
+
+	| newDbgPosition lastUserPosition |
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user position (test running should not modify user changes)"
+	lastUserPosition := self settings lastKnownPosition.
+	newDbgPosition := 500 @ 100.
+	"Programmatically change the position of debugger window"
+	dbg window window position: newDbgPosition.
+	"New position is saved when the window is closed"
+	dbg close.
+	self tearDown.
+
+	"Now we open the debugger again"
+	self setUp.
+	dbg := (self debuggerOn: session) open.
+	"Check the debugger window is set to the new position"
+	self assert: dbg window window position equals: newDbgPosition.
+	"Set back the last position regardless of this test"
+	dbg window window position: lastUserPosition.
+
+]
+
```
1) I feel like each of these tests should be split into two smaller tests:
- one test that checks that the value `lastKnownExtent` (resp. `lastKnownPosition`) from `StDebuggerSettings` is changed when a debugger window is closed
- one test that when opening a new debugger, the extent (resp. position) of its window is the one registered in `StDebuggerSettings`

2) Also, it is really weird to call `#setUp` and `#tearDown` in a test. 
I think it breaks a lot of things that forces you to add weird `ifNotNil:` checks.
If the debugger needs to be cleared, then you could extract  from the `#tearDown` a method that clears the debugger and that you could call in your tests, instead of calling `tearDown`  and `setUp`

3) When setting back the values of  `lastKnownExtent` (resp. `lastKnownPosition`), I think that you should use the setters of `StDebugger` settings in an `ensure:` block

hernanmd replied:
I agree with adding two tests. But please note that they are not the same. Opening the debugger window and then programmatically changing its position ensures that we test real user interaction with the window positioning. So, the position change should be updated **from the interaction**, not just from the global, and it should match the position set in the test. 

If you do not open an actual debugger window, you check if a global variable was updated accordingly.

Regarding the manual calling of `tearDown` and `setUp`, did you try to execute the test without sending `tearDown` and `setUp`? When I tried, many things started to fail in the debugger's cleanup.

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,56 @@ StDebuggerTest >> testNewDebuggerContextFor [
 ]
 
+{ #category : 'tests - context inspector' }
+StDebuggerTest >> testPreserveLastKnownExtent [
+
+	| newDbgExtent lastUserExtent |
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user extent (test running should not modify user changes)"
+	lastUserExtent := self settings lastKnownExtent.
+	newDbgExtent := 100 @ 200.
+	"Programmatically change the extent of debugger window"
+	dbg window window extent: newDbgExtent.
+	"New extent is saved when the window is closed"
+	dbg close.
+	self tearDown.
+
+	"Now we open the debugger again"
+	self setUp.
+	dbg := (self debuggerOn: session) open.
+	"Check the debugger window is set to the new extent"
+	self assert: dbg window window extent equals: newDbgExtent.
+	"Set back the last extent regardless of this test"
+	dbg window window extent: lastUserExtent.
+
+]
```
this test is failing

hernanmd replied:
Which Pharo version do you use?

Using the latest Pharo 13.0.0 here, and it works:
Build information: Pharo-13.0.0+SNAPSHOT.build.124.sha.d1dbcc0440c7d27e9a05e6ea2a659f5ddb41a484 (64 Bit)

![Screenshot 2024-07-15 at 15 53 55](https://github.com/user-attachments/assets/f7233034-2500-4fe5-aed8-61540f53188b)

adri09070 replied:
the last P13 build (look at the CI)

hernanmd replied:
I see, I don't understand why is it failing in the CI.

I would rewrite the test following your suggestions, although I think it doesn't reflect the actual user interaction.

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,56 @@ StDebuggerTest >> testNewDebuggerContextFor [
+	lastUserExtent := self settings lastKnownExtent.
+	newDbgExtent := 100 @ 200.
+	"Programmatically change the extent of debugger window"
+	dbg window window extent: newDbgExtent.
+	"New extent is saved when the window is closed"
+	dbg close.
+	self tearDown.
+
+	"Now we open the debugger again"
+	self setUp.
+	dbg := (self debuggerOn: session) open.
+	"Check the debugger window is set to the new extent"
+	self assert: dbg window window extent equals: newDbgExtent.
+	"Set back the last extent regardless of this test"
+	dbg window window extent: lastUserExtent.
+
+]
+
+{ #category : 'tests - context inspector' }
+StDebuggerTest >> testPreserveLastKnownPosition [
+
+	| newDbgPosition lastUserPosition |
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user position (test running should not modify user changes)"
+	lastUserPosition := self settings lastKnownPosition.
+	newDbgPosition := 500 @ 100.
+	"Programmatically change the position of debugger window"
+	dbg window window position: newDbgPosition.
+	"New position is saved when the window is closed"
+	dbg close.
+	self tearDown.
+
+	"Now we open the debugger again"
+	self setUp.
+	dbg := (self debuggerOn: session) open.
+	"Check the debugger window is set to the new position"
+	self assert: dbg window window position equals: newDbgPosition.
+	"Set back the last position regardless of this test"
+	dbg window window position: lastUserPosition.
```
maybe we should add tests that test the behavior when `StDebuggerSettings class>>#PreserveWindowPositionAndExtent` is false

hernanmd committed:
Merge branch 'pharo-spec:Pharo13' into p13_allow_last_debugger_position_extent

Commit:
Updated test that checks that the value lastKnownExtent from StDebuggerSettings is changed when a debugger window is closed.
Updated test that checks that the value lastKnownPosition from StDebuggerSettings is changed when a debugger window is closed.
Removed setUp and tearDown sends from tests.
Add tests the for StDebuggerSettings class>>#preserveWindowPositionAndExtent is false and true

Commit:
Add two new test to check that the opened window is opening in the saved position and extent.

hernanmd committed:
Merge branch 'Pharo13' into p13_allow_last_debugger_position_extent

hernanmd commented:
Guys, yesterday this PR was green and today I've seen some conflicts so it would be nice to have it integrated on time before other PR's make it obsolete, otherwise, I have to work on this forever.

Now I'm checking this again because StDebugger initializeWindow: was changed somewhere.

hernanmd commented:
The failing test is unrelated.

hernanmd commented:
> Guys, yesterday this PR was green and today I've seen some conflicts so it would be nice to have it integrated on time before other PR's make it obsolete, otherwise, I have to work on this forever.
> 
> Now I'm checking this again because StDebugger initializeWindow: was changed somewhere.

Done. Could you have a look @StevenCostiou ?

StevenCostiou commented:
Perhaps tomorrow, I am not available today.
But worry not, the debugger code does not change that often, that is just bad luck.

hernanmd committed:
Merge branch 'Pharo13' into p13_allow_last_debugger_position_extent

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+
+	| lastUserPreservation |
+
+	lastUserPreservation := self settings preserveWindowPositionAndExtent.
```
this should be done in `setUp`. If the test fails, then this settings won't be set to how it used to be

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+	dbg close.
+
+	"Restore the original preservation setting"		
+	self settings preserveWindowPositionAndExtent: lastUserPreservation.
```
this should be done in `tearDown`, for the same reasons

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+
+	| lastUserPreservation |
+
+	lastUserPreservation := self settings preserveWindowPositionAndExtent.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+	dbg close.
+
+	"Restore the original preservation setting"		
+	self settings preserveWindowPositionAndExtent: lastUserPreservation.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+
+	| newDbgExtent lastUserPreservation |
+
+	lastUserPreservation := self settings preserveWindowPositionAndExtent.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+		equals: newDbgExtent.
+
+	"Restore the original preservation setting"		
+	self settings preserveWindowPositionAndExtent: lastUserPreservation.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user extent (test running should not modify user changes)"
+	lastUserExtent := self settings lastKnownExtent.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+		equals: newDbgExtent.
+
+	"Restore the original extent"		
+	self settings lastKnownExtent: lastUserExtent.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+
+	| newDbgPosition lastUserPreservation |
+
+	lastUserPreservation := self settings preserveWindowPositionAndExtent.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+		equals: newDbgPosition.
+
+	"Restore the original preservation setting"		
+	self settings preserveWindowPositionAndExtent: lastUserPreservation.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user position (test running should not modify user changes)"
+	lastUserPosition := self settings lastKnownPosition.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -752,6 +759,136 @@ StDebuggerTest >> testNewDebuggerContextFor [
+		equals: self settings lastKnownPosition.
+		
+	"Restore the original position"		
+	self settings lastKnownPosition: lastUserPosition.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -1216,3 +1353,47 @@ StDebuggerTest >> testUpdateSourceCodeFor [
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user position (test running should not modify user changes)"
+	lastUserExtent := self settings lastKnownExtent.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -1216,3 +1353,47 @@ StDebuggerTest >> testUpdateSourceCodeFor [
+		equals: newDbgExtent.
+
+	"Restore the original extent"		
+	self settings lastKnownExtent: lastUserExtent.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -1216,3 +1353,47 @@ StDebuggerTest >> testUpdateSourceCodeFor [
+
+	dbg := (self debuggerOn: session) open.
+	"First we save the user position (test running should not modify user changes)"
+	lastUserPosition := self settings lastKnownPosition.
```
should be done in `setUp`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -1216,3 +1353,47 @@ StDebuggerTest >> testUpdateSourceCodeFor [
+		equals: newDbgPosition.
+		
+	"Restore the original position"		
+	self settings lastKnownPosition: lastUserPosition.
```
should be done in `tearDown`

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -121,10 +127,11 @@ StDebuggerTest >> tearDown [
+		oldMethodToSourceDictionary	associationsDo: [ :assoc |
+ 			assoc key methodClass
+ 				compile: assoc value
+ 				classified: assoc key protocolName ] ].
```
this `ifNil:` call  should really be unnecessary

adri09070 commented on src/NewTools-Debugger-Tests/StDebuggerTest.class.st:
```diff
@@ -103,7 +109,7 @@ StDebuggerTest >> subscriptionsForStDebuggerExtensionActivationToggle [
 StDebuggerTest >> tearDown [
 
-	session clear.
+	session ifNotNil: [ session clear ].
```
this `ifNil:` call should really be unnecessary

adri09070 reviewed (changes requested):
Your test fails randomly and I can't exactly explain why but here are things that should be changed in your tests and that will make your test pass everytime

hernanmd committed:
Merge branch 'pharo-spec:Pharo13' into p13_allow_last_debugger_position_extent

Commit:
Implemented refactoring suggestions by @adri09070

StevenCostiou commented:
For me now there are two problems left:
- what about those nil checks? can we do something?
- tests are all red (I didn't check what's happening)
If those are fixed, can we agree to merge?

adri09070 commented:
Concerning the failing tests, they are not related

hernanmd committed:
Merge branch 'pharo-spec:Pharo13' into p13_allow_last_debugger_position_extent

hernanmd committed:
Merge branch 'pharo-spec:Pharo13' into p13_allow_last_debugger_position_extent

Commit:
Remove possibly unneeded nil checks in tearDown method.

hernanmd commented:
> For me now there are two problems left:
> 
>     * what about those nil checks? can we do something?
> 
>     * tests are all red (I didn't check what's happening)
>       If those are fixed, can we agree to merge?

I removed the ifNil: checks since I need to work on other issues.
